### PR TITLE
Development Cloning & Various small changes

### PR DESCRIPTION
**Pull request by Cortesj**
Adds 4 new flags to the **BUILD** command pertaining to repository cloning. There is checks to see if remote repository is a valid repository and will clone the default RunestoneInteractive remote repositories if a error such as 'git error 128 ' or any other subprocess error is passed into script. 

Fixes a issue with 1.py template located within docker_tools.py. There was a syntax error with the f string that uses RUNESTONE_PATH, import error of missing import for function ENVIRON and missing "_" from settings variable. 

Added to ```docker_tools.py build [options]```
- ```-crs / --clone-rs``` | Clone RunestoneServer with "X" Username
- ```-cbs / --clone-bs``` | Clone BookServer with "X" Username
- ```-crc / --clone-rc``` | Clone RunestoneComponents with "X" Username
- ```-c / --clone-all``` | Clone All repositories with "X" Username

Related to my issue #1775

Would @bjones1 and @bnmnetp  be able to review this to see if I need to potentially make some changes or potentially relocated to some other script.

### CODE REVIEW

File: docker/docker_tools.py
Comment (bjones1) [inline docker/docker_tools.py:279]: Also check that the other clone flags aren't set. If they are, issue a warning that their values are ignored, or even quit with an error.

File: docker/docker_tools.py
Comment (bjones1) [inline docker/docker_tools.py:292]: You can use `xqt("GIT_TERMINAL_PROMPT=0...` instead (no need to export).

File: docker/docker_tools.py
Comment (bjones1) [inline docker/docker_tools.py:293]: I suggest omitting this -- if the repo clone fails, exit with an error. Otherwise, the user won't see the difference in all the output produced by a build.

File: docker/docker_tools.py
Comment (bjones1) [reply docker/docker_tools.py:292]: Also, use an f-string instead -- that's easier to read.

File: docker/docker_tools.py
Comment (bjones1) [inline docker/docker_tools.py:364]: Same comment -- exit with error.

File: docker/docker_tools.py
Comment (bjones1) [inline docker/docker_tools.py:363]: Same comment -- no `export`.

File: docker/docker_tools.py
Comment (bjones1) [inline docker/docker_tools.py:374]: Same comments.

File: docker/docker_tools.py
Comment (bjones1) [inline docker/docker_tools.py:396]: Why do this? I'd suggest keeping all the flags, just in case we want to refer to them later.

**Review by bjones1 — CHANGES_REQUESTED** [review]

File: docker/docker_tools.py
Comment (Cortesj) [reply docker/docker_tools.py:279]: I'm confused because default repositories is "RunestoneInteractive" and if the ```-c / --clone-all``` flag is even touched then it will set all the  repositories to attempt cloning with specified username. Are you saying that I should implement a check for each repository that will error out/exit if it cannot clone from specified user's remote repositories?

File: docker/docker_tools.py
Comment (Cortesj) [reply docker/docker_tools.py:396]: The only reason I removed the clone flags from being passed though is due to issues with the actual building of the docker containers. I might be able to replicate the issue.

File: docker/docker_tools.py
Comment (Cortesj) [reply docker/docker_tools.py:187]: I never heard of black till now and it seems like such a sweet little program! I used it in latest commit to PR.

File: docker/docker_tools.py
Comment (Cortesj) [reply docker/docker_tools.py:292]: Must have forgot f-string is a thing again... Anyway, I changed them to f-strings like you suggested and it appears that export is necessary to temporarily disable the Git Prompt to pass-though the error. I attempted to execute it without ```export``` but it would act as if it was never executed. I did noticed that the export only affects that session of script execution so it isn't exporting to the host system.

File: docker/docker_tools.py
Comment (Cortesj) [reply docker/docker_tools.py:293]: I decided to change the except to perform a "sys.exit()" with a message stating there was an error cloning the system a remote repository with specific username.

File: docker/docker_tools.py
Comment (Cortesj) [reply docker/docker_tools.py:364]: Changed exit to error out with message.

File: docker/docker_tools.py
Comment (Cortesj) [reply docker/docker_tools.py]: Previously commented about ```export``` command being used for cloning.

File: docker/docker_tools.py
Comment (Cortesj) [reply docker/docker_tools.py:374]: Noted and changed in latest Commit.

**Comment by Cortesj** [comment]
Latest commit 8886946aa1cb8368d6708bf891b1827f88f8fcba I did the majority of the suggestions about system error exits, f-strings, and  etc. Waiting on some more suggestions and discussion about ```export``` and passing all arguments to the containers. I knew you had some questions about not passing the clone flags.

File: docker/docker_tools.py
Comment (bjones1) [reply docker/docker_tools.py:187]: Great!

File: docker/docker_tools.py
Comment (bjones1) [reply docker/docker_tools.py:279]: Here, I'm asking for a CLI consistency check: it doesn't make sense (to me) to specify both `--clone-all=my_repo` while also specifying `--clone_rc=my_other_repo`. I can see three solutions:

1. In the help text, state that value supplied in `--clone-rs`, `--clone-bs`, and `--clone-rc` will override the value supplied in `--clone-all`. Also issue a warning when this happens.
2. Issue an error, telling user that the same value was specified twice.

I'll leave it to you to decide which is best.

On your second point: yes, you definitely want to error if a git clone from the specified repo fails.

File: docker/docker_tools.py
Comment (bjones1) [reply docker/docker_tools.py:293]: Good. I think that `sys.exit()` requires an int, though? You can use [click.echo](https://click.palletsprojects.com/en/8.0.x/quickstart/#echoing) to display a message.

File: docker/docker_tools.py
Comment (bjones1) [reply docker/docker_tools.py:394]: OK, let me know.

**Review by bjones1 — CHANGES_REQUESTED** [review]

File: docker/docker_tools.py
Comment (Cortesj) [reply docker/docker_tools.py:293]: ```sys.exit()``` has an optional argument. If nothing is specified within the function or ```None``` value is passed then the default exit code is 0. I know you can pass a int between 0-127 to refer to different error types. I passed a string object which prints out to STDERR (Terminal) and returns an exit code 1. 

It worked during testing by providing it non-existing usernames/repositories to see if ERROR reported out correctly.

Reference:
[Python sys module Documentation](https://docs.python.org/3/library/sys.html)

File: docker/docker_tools.py
Comment (Cortesj) [reply docker/docker_tools.py:279]: > Here, I'm asking for a CLI consistency check: it doesn't make sense (to me) to specify both --clone-all=my_repo while also specifying --clone_rc=my_other_repo. I can see three solutions:

Okay, I think I see your confusion. You are only meant to specify ```--clone-all``` if you wish to clone all the repositories with your remote repositories. Hypothetically, if you only want to clone BookServer from your own remote repository then you would specify ```--clone-bs```. 

I think I will be more specific within the ```--help``` stating the outcome of using ```--clone-all``` in tandem with any other clone flag. I could also provide some examples within ```--help``` if that would help with increasing the thought behind its functionality.
<br><br/>
> On your second point: yes, you definitely want to error if a git clone from the specified repo fails.

Addressed this with ```sys.exit()``` on another comment.

File: docker/docker_tools.py
Comment (bjones1) [reply docker/docker_tools.py]: OK, sounds good.

File: docker/docker_tools.py
Comment (bjones1) [reply docker/docker_tools.py:293]: Thanks, I learned something (sys.exit(string))!

**Comment by bjones1** [comment]
This is close -- AFAIK, the only remaining items are:

1. Add clarification / help messages on what happens if the user specifies `--clone-all` and `--clone-xx`
2. Clean up the arg pass to Docker.

**Comment by Cortesj** [comment]
Created a few new commits to address the suggestions involving cloning clarification and help messages

1. Commit 9e1c34ae7e9361b70d46ca267b4d471bbc08e98d
* Added a WARNING for using the ```--clone-all``` flag within help description
* Changed the METAVAR to display <USERNAME> rather than TEXT
Screenshot from 2021-11-24 00-15-54

* Removed the argument pass filter since after testing I could not replicate previous issue
* Initiate 20 second countdown on how to abort script if ```--clone-all``` flag is set
Screenshot from 2021-11-24 00-09-36

2. Commit e4e0bc4238cab6b235086374fef688117eca66ff
Clean up of code with ```Black```

File: docker/docker_tools.py
Comment (Cortesj) [reply docker/docker_tools.py]: Addressed this with commit 9e1c34ae7e9361b70d46ca267b4d471bbc08e98d

File: docker/docker_tools.py
Comment (Cortesj) [reply docker/docker_tools.py]: After some testing, I never was able to replicate the error. I currently tested it with multiple combinations of the ```--clone-XX``` flag.

File: docker/docker_tools.py
Comment (bjones1) [inline docker/docker_tools.py:298]: I'd omit this. A warning in red is enough.

**Review by bjones1 — CHANGES_REQUESTED** [review]
Mostly there! Just remove the 20 second delay.

File: docker/docker_tools.py
Comment (Cortesj) [reply docker/docker_tools.py]: Commit 4bccb7c024d88799ceed7007694838722406baf1 removes delay.

**Review by bjones1 — APPROVED** [review]
Look good!

**Comment by Cortesj** [comment]
@bjones1 It seems my pull request failed to merge due to the server tests. I know that within Docker_build the error comes to not setting the BookServer clone default to "bjones1". I am just confused on why ubunutu_build failed.

**Comment by bjones1** [comment]
It looks like Github was down/not working well which caused the `ubuntu_build` failures. Just make a small change and push and it will re-run the tests. They should work better after a retry.

**Comment by Cortesj** [comment]
@bjones1 Okay, I made a few small changes mostly some misplaced variable names and the default for BookServer since it isn't officially forked by RunestoneInteractive yet. It says there is a merge conflict only those with write access can fix but I am not sure why it got flagged as merge conflict since it simply replaces some current code for RunestoneComponents with what I wrote.

**Pull request merged**
